### ISSUE TITLE
I can't use Camera2

### ISSUE THREAD

**Issue by LutfiTekin** [pedroSG94/RootEncoder#162]
RtmpCamera1 is working fine but when I switch to using RtmpCamera2 it's all black screen and I get these logs:

```
04-23 21:43:52.940 19497-20486/****** I/Camera2ApiManager: camera disconnected

04-23 21:43:52.987 19497-20507/****** I/Camera2ApiManager: camera opened

04-23 21:43:52.987 19497-20507/****** E/Camera2ApiManager: configuration failed
```

I don't know the benefits of using camera2 api. If it doesn't matter I can continue using RtmpCamera1 but if it has advantages then I need your help.

**Comment by pedroSG94** [comment]
Normally camera2 has more performance and more resolution to open (camera1 is cap to 1080p because in stream you need get frames from preview)

Can you tell me device model and Android API?

**Comment by LutfiTekin** [comment]
Xperia XZ android 8.0

**Comment by pedroSG94** [comment]
Make sure that you are using a resolution of listed in this method:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/base/Camera2Base.java#L466

Do you have this permissions in manifest?:
```xml
<uses-feature android:name="android.hardware.camera" />
<uses-feature android:name="android.hardware.camera2.full" />
<uses-feature android:name="android.hardware.camera2.autofocus" />
```

Also add this method in this line to show camera2 level supported by your device:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java#L287

```java
private void showLevelSupported(CameraCharacteristics c) {
    int deviceLevel = c.get(CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL);
    if (deviceLevel == CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LEGACY) {
      Log.i(TAG, "Level supported: legacy");
    } else if (deviceLevel == CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_3) {
      Log.i(TAG, "Level supported: level 3");
    } else if (deviceLevel == CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_FULL){
      Log.i(TAG, "Level supported: full");
    } else if (deviceLevel == CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LIMITED){
      Log.i(TAG, "Level supported: limited");
    }
  }
```

**Comment by LutfiTekin** [comment]
Yes I have those in my manifest.
And for adding showLevelSupported to that line honestly I don't know how to edit an external library. Can you show me how I can add this lib locally?

**Comment by pedroSG94** [comment]
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/wiki/Add-library-to-your-project

Other way if you only want test is download my project with app example and modify in that project(To download you can clone with git or download zip).

**Comment by LutfiTekin** [comment]
Ok I am on it now

**Comment by LutfiTekin** [comment]
I ran the but it uses camera1

`04-24 15:48:55.343 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 1600X1200, not supported for encoder
04-24 15:48:55.718 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 1600X1200, not supported for encoder
04-24 15:48:56.586 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 640X480
04-24 15:49:21.317 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 1600X1200, not supported for encoder
04-24 15:49:21.590 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 1600X1200, not supported for encoder
04-24 15:49:22.238 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 640X480
04-24 15:49:25.673 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 1600X1200, not supported for encoder
04-24 15:49:25.927 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 1600X1200, not supported for encoder
04-24 15:49:26.389 22060-22060/com.pedro.rtpstreamer I/Camera1ApiManager: 640X480`

**Comment by LutfiTekin** [comment]
I found a snippet from [stack overflow ](https://stackoverflow.com/a/40069240/3742074)

```
04-24 15:53:25.475 23048-23048/? I/CameraManagerGlobal: Connecting to camera service
04-24 15:53:25.493 23048-23048/? D/Camera2Control: Camera 0 has FULL Camera2 support
04-24 15:53:25.494 23048-23048/? D/Camera2Control: Camera 1 has LIMITED Camera2 support
```

**Comment by LutfiTekin** [comment]
> Make sure that you are using a resolution of listed in this method:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/base/Camera2Base.java#L466


I didn't get this part I am using it like this, I don't set a resolution

```
if(broadcastCam.prepareAudio() && broadcastCam.prepareVideo()){
                broadcastCam.startPreview(1)
                broadcastCam.setAuthorization(LOGIN,PASS)
                broadcastCam.startStream(RTMPURL)
   
            }
```

**Comment by pedroSG94** [comment]
If you use prepareVideo() you select 640x480 by default but no all device support this resolution, anyway if it work with camera1, camera2 should has same resolution supported.
Also camera2 support hardware should be ok. I need investigate more about it. If you can try debug you too because found a phone with that bug could be hard.
You can try open camera 2 with this example that has code more or less like my library:
https://github.com/pedroSG94/camera2-and-samsung-camera-example

**Comment by LutfiTekin** [comment]
Testing this but still getting black screen
```
broadcastCam = RtmpCamera2(broadcastSurface,connectionChecker)
            if(broadcastCam.prepareAudio() && broadcastCam.prepareVideo(480,720,30,5000,false,180)){
                broadcastCam.startPreview(1)
                broadcastCam.setAuthorization(LOGIN,PASS)
                broadcastCam.startStream(RTMPURL)
            }
```

getting these logs

```
04-24 17:25:30.999 21704-21704/******** W/******: type=1400 audit(0.0:3152): avc: denied { read } for name="u:object_r:camera_prop:s0" dev="tmpfs" ino=15633 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:camera_prop:s0 tclass=file permissive=0 ppid=1019 pcomm="main" pgid=1019 pgcomm="main"
04-24 17:25:31.010 21704-21704/******** E/libc: Access denied finding property "persist.camera.cfa.packagelist"
04-24 17:25:31.152 21704-22269/******** I/Camera2ApiManager: camera disconnected
04-24 17:25:31.167 21704-22296/******** W/CameraDevice-JV-0: Stream configuration failed due to: createStream:513: Camera 0: Surface already has a stream created for it (ID 0)
04-24 17:25:31.168 21704-22296/******** E/CameraCaptureSession: Session 0: Failed to create capture session; configuration failed
04-24 17:25:31.168 21704-22296/******** I/Camera2ApiManager: camera opened
04-24 17:25:31.168 21704-22296/******** E/Camera2ApiManager: configuration failed
```

**Comment by pedroSG94** [comment]
I was looking info about it but no result for now.
Notify me if you find any solution or more info about this error.

**Comment by pedroSG94** [comment]
If you can try open this google example of camera2 in your device that no work:
https://github.com/googlesamples/android-Camera2Basic
If this work we need change Camera2ApiManager to fix it using that example. If no work write a issue in that project and I will try find more info about it to solve

**Comment by LutfiTekin** [comment]
Ok, I will try it thanks

**Comment by ramkishoreprajapati** [comment]
finally I resolved this issue in one plus 5T (Android 8.1.0).

Problem in this line
captureRequestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE);

Use this **TEMPLATE_PREVIEW** instead of **TEMPLATE_STILL_CAPTURE**

example
captureRequestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);

**Comment by pedroSG94** [comment]
This library actually use TEMPLATE_PREVIEW:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java#L152

Confirm me that this line no produce error with your device:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java#L165
